core/src/main/java/hudson/PluginManager.java

```diff
@@ -1341,41 +1343,55 @@ public static boolean isNonMetaLabel(String label) {
     
     @Restricted(NoExternalUse.class)
     public HttpResponse doPluginsSearch(@QueryParameter String query, @QueryParameter Integer limit) {
+        final String _query = query.toLowerCase(Locale.ROOT);
+        final boolean isWarningsQuery = _query.contains("warning:");
         List<JSONObject> plugins = new ArrayList<>();
         for (UpdateSite site : Jenkins.get().getUpdateCenter().getSiteList()) {
             List<JSONObject> sitePlugins = site.getAvailables().stream()
                 .filter(plugin -> {
-                    if (StringUtils.isBlank(query)) {
+                    // preserve an existing feature for the "warning:" query
+                    if (isWarningsQuery && plugin.hasWarnings()) {
                         return true;
                     }
-                    return StringUtils.containsIgnoreCase(plugin.name, query) ||
-                        StringUtils.containsIgnoreCase(plugin.title, query) ||
-                        StringUtils.containsIgnoreCase(plugin.excerpt, query) ||
-                        plugin.hasCategory(query) ||
-                        plugin.getCategoriesStream()
-                            .map(UpdateCenter::getCategoryDisplayName)
-                            .anyMatch(category -> StringUtils.containsIgnoreCase(category, query)) ||
-                        plugin.hasWarnings() && query.equalsIgnoreCase("warning:");
-                })
-                .limit(Math.max(limit - plugins.size(), 1))
-                .sorted((o1, o2) -> {
-                    String o1DisplayName = o1.getDisplayName();
-                    if (o1.name.equalsIgnoreCase(query) ||
-                        o1DisplayName.equalsIgnoreCase(query)) {
-                        return -1;
-                    }
-                    String o2DisplayName = o2.getDisplayName();
-                    if (o2.name.equalsIgnoreCase(query) || o2DisplayName.equalsIgnoreCase(query)) {
-                        return 1;
+                    // if the query is a category, return plugins in that category
+                    if (plugin.hasCategory(_query)) {
+                        return true;
                     }
-                    if (o1.name.equals(o2.name)) {
-                        return 0;
+                    // otherwise we need to find a specific plugin based on the query
+                    // for short queries, we want plugins whose title or name starts with the query
+                    // for longer, multi-word queries, we want any part of the title to match the query words
+                    final String _title = plugin.title.toLowerCase(Locale.ROOT);
```

Comment:
💡 If we want to keep consistency within Jenkins core, one should use `Locale.ENGLISH`. Correct or wrong, the idea is to have a single usage within the project, not something author dependent that will make maintenance more difficult.

- Locale.ROOT: 1 occurrence in a test class
- Locale.ENGLISH: 53 matches
- Locale.US: 8 matches
- Locale.getDefault(): 8 matches

Reply:
Will do. The plugin name (ID) and title are in English, so Locale.ROOT isn't needed here.